src/Common/Vault.cpp

```diff
@@ -0,0 +1,92 @@
+#include <string>
+#include <Common/Exception.h>
+#include <Common/Vault.h>
+
+#include <IO/ConnectionTimeouts.h>
+#include <IO/HTTPCommon.h>
+#include <IO/ReadBuffer.h>
+#include <IO/ReadWriteBufferFromHTTP.h>
+#include <Poco/JSON/Array.h>
+#include <Poco/JSON/Parser.h>
+#include <Poco/URI.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+}
+
+Vault & Vault::instance()
+{
+    static Vault ret;
+    return ret;
+}
+
+void Vault::load(const Poco::Util::AbstractConfiguration & config, const String & config_prefix, ContextPtr context_)
+{
+    reset();
+
+    context = context_;
+
+    url = config.getString(config_prefix + ".url", "");
+
+    if (url.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "url is not given for vault.");
+
+    token = config.getString(config_prefix + ".token", "");
+
+    if (token.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "token is not given for vault.");
+
+    loaded = true;
+}
+
+String Vault::readSecret(const String & secret, const String & key)
+{
+    LOG_DEBUG(log, "readSecret {} {}", secret, key);
+
+    DB::HTTPHeaderEntries headers;
+    headers.emplace_back("X-Vault-Token", token);
+    Poco::URI uri = Poco::URI(fmt::format("{}/v1/secret/data/{}", url, secret));
+    Poco::Net::HTTPBasicCredentials credentials{};
+
+    std::string json_str;
+
+    try
+    {
+        auto wb = DB::BuilderRWBufferFromHTTP(uri)
+                      .withConnectionGroup(DB::HTTPConnectionGroupType::HTTP)
+                      .withMethod(Poco::Net::HTTPRequest::HTTP_GET)
+                      .withTimeouts(DB::ConnectionTimeouts::getHTTPTimeouts(context->getSettingsRef(), context->getServerSettings()))
+                      .withSkipNotFound(false)
+                      .withHeaders(headers)
+                      .create(credentials);
+
+        readJSONObjectPossiblyInvalid(json_str, *wb);
```

Comment:
Is it ok to ignore return code?

Reply:
Right, return type is void unless 'bool' is explicitly specified.
If you are sure that exception is handled (anywhere) in a reasonable way, that's cool.

Reply:
You may search ClickHouse source code for readJSONObjectPossiblyInvalid with return type bool. It's used with void in 99% of cases. I could not find it with bool.